README.md
# Quiron

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![Twitter](https://img.shields.io/twitter/follow/luctstt.svg?label=Follow&style=social)](https://twitter.com/iluzioDev)

---

<video width="100%" controls="controls" src="https://user-images.githubusercontent.com/45295283/211917624-f9b2433e-9d37-4c20-97c5-2fb3fb93b326.mp4">Your browser does not support the HTML5 Video element.</video>

## Introduction

Quiron is a Python package that provides hand detection, gesture recognition, and finger counting, so developers can easily integrate hand tracking and gesture recognition functionality into their applications, allowing for more natural and intuitive user interactions.

Quiron leverages the powerful [Mediapipe framework](https://mediapipe.dev/), which uses machine learning models to accurately detect hands in images and video streams. Once hands are detected, Quiron provides a range of functionalities for recognizing different gestures, such as thumbs up or peace sign, as well as counting the number of fingers that are raised.

## Features

- Hand Recognition
- Hand Movement Detection
- Hand Gesture Recognition
- Finger Counting

## Installation

```console
pip install quiron
```

## API

### Hand Recognition

```{eval-rst}
.. autofunction:: quiron.count_fingers
```

```{eval-rst}
.. autofunction:: quiron.detect_gestures
```

```{eval-rst}
.. autofunction:: quiron.gestures.distance
```

```{eval-rst}
.. autofunction:: quiron.gestures.raised
```

```{eval-rst}
.. autofunction:: quiron.gestures.sideways
```

## Maintainers

<table>
  <tr>
    <td align="center"><a href="https://github.com/iluzioDev"><img src="https://avatars.githubusercontent.com/u/45295283?v=4" width="100px;" alt="IluzioDev"/><br /><sub><b>IluzioDev</b></sub></a><br />💻</td>
  </tr>
</table>

## License ⚖️

Distributed under the MIT License. [Click here](LICENSE.md) for more information.

---

<div align="center">
	<b>
		<a href="https://www.npmjs.com/package/get-good-readme">File generated with get-good-readme module</a>
	</b>
</div>


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2023-04-06

### Changed
  
- Separated the functions and constants into different files, ```gestures.py``` and ```constants.py```, respectively.


quiron.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Dec 20 2022

@author: iluzioDev

This script implements Hand Recognition using Mediapipe and OpenCV.
Includes features like:
  - Hand Recognition
  - Hand Movement Detection
  - Hand Gesture Recognition
  - Finger Counting
"""
import cv2
import mediapipe

import modules.constants as constants
import modules.gestures as gestures

# Flags to detect Hand movement
moves = prev_thumb = prev_index = prev_middle = prev_ring = prev_pinky = False

def count_fingers(landmark, label):
  """
  Counts the number of fingers up in a hand
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
    
  Returns:
    int: Number of fingers up
  """
  count = 0
  for i in range(5):
    if gestures.raised(landmark, label, constants.TIPS[i]):
      count += 1
  return count

def detect_gestures(landmark, label):
  """
  Detects the gesture of a hand

  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
  
  Returns:
    str or bool: The gesture if it is detected, False otherwise
  """
  if gestures.thumb_up(landmark, label):
    return "Thumbs Up! :D"
  if gestures.thumb_down(landmark, label):
    return "Thumbs Down... :("
  if gestures.peace(landmark, label):
    return "Peace!"
  if gestures.rock_n_roll(landmark, label):
    return "Rock'n'Roll!"
  if gestures.surf(landmark, label):
    return "Surf's up!"
  if gestures.ok(landmark, label):
    return "Everything is OK"
  if gestures.loser(landmark, label):
    return "Loser >:D"
  return False

def moving(landmark):
  """
  Detects if the hand is moving or not calculating the distance in a margin between the positions
  of 2 instances through time of the hand

  Args:
    landmark: Landmark of one hand
  
  Returns:
    bool: True if the hand is moving, False otherwise
  """
  global moves, prev_thumb, prev_index, prev_middle, prev_ring, prev_pinky
  if not prev_thumb:
    prev_thumb = landmark[constants.TIPS[0]]
    prev_index = landmark[constants.TIPS[1]]
    prev_middle = landmark[constants.TIPS[2]]
    prev_ring = landmark[constants.TIPS[3]]
    prev_pinky = landmark[constants.TIPS[4]]

  d0 = gestures.distance(prev_thumb, landmark[constants.TIPS[0]])
  d1 = gestures.distance(prev_index, landmark[constants.TIPS[1]])
  d2 = gestures.distance(prev_middle, landmark[constants.TIPS[2]])
  d3 = gestures.distance(prev_ring, landmark[constants.TIPS[3]])
  d4 = gestures.distance(prev_pinky, landmark[constants.TIPS[4]])

  if (d0 + d1 + d2 + d3 + d4) > constants.MARGIN:
    moves = True
  else:
    moves = False

  prev_thumb = landmark[constants.TIPS[0]]
  prev_index = landmark[constants.TIPS[1]]
  prev_middle = landmark[constants.TIPS[2]]
  prev_ring = landmark[constants.TIPS[3]]
  prev_pinky = landmark[constants.TIPS[4]]

  return moves

def main():
  """
  Main function of the program. Starts the WebCam Capture and uses the features implemented
  
  Returns:
    None
  """
  mp_drawing = mediapipe.solutions.drawing_utils
  mp_drawing_styles = mediapipe.solutions.drawing_styles
  mp_hands = mediapipe.solutions.hands

  # Starts the WebCam Capture
  capture = cv2.VideoCapture(0)
  with mp_hands.Hands(model_complexity=0, min_detection_confidence=0.5, min_tracking_confidence=0.5) as hands:
    while capture.isOpened():
      success, frame = capture.read()
      if not success:
        print("Ignoring empty camera frame.")
        # If loading a video, use 'break' instead of 'continue'.
        continue

      # Flip the frame horizontally for a selfie-view display.
      frame = cv2.flip(frame, 1)

      # To improve performance, optionally mark the frame as not writeable to
      # pass by reference.
      frame.flags.writeable = False
      frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
      results = hands.process(frame)

      # Draw the hand annotations on the frame.
      frame.flags.writeable = True
      frame = cv2.cvtColor(frame, cv2.COLOR_RGB2BGR)

      # height, width and depth (RGB=3) of frame
      (height, width, depth) = frame.shape

      # finger_count of gestures.raised fingers
      finger_count = 0

      # If hands are detected
      if results.multi_hand_landmarks:
        for hand_landmarks in results.multi_hand_landmarks:
          # Index Position and Handedness (left hand or rigth hand)
          index = results.multi_hand_landmarks.index(hand_landmarks)
          label = results.multi_handedness[index].classification[0].label

          # Drawing within frame of the hand landmark with its connections and mediapipe default style
          mp_drawing.draw_landmarks(
              frame,
              hand_landmarks,
              mp_hands.HAND_CONNECTIONS,
              mp_drawing_styles.get_default_hand_landmarks_style(),
              mp_drawing_styles.get_default_hand_connections_style())

          # Detect fingers
          landmark = hand_landmarks.landmark
          finger_count += count_fingers(landmark, label)

          # Detect Gestures
          if detect_gestures(landmark, label):
            cv2.rectangle(frame, (int(0.2 * width), int(0.825 * height)),
                          (int(0.975 * width), int(0.975 * height)), constants.WHITE, -1)
            cv2.putText(frame, detect_gestures(landmark, label), (int(0.225 * width), int(0.925 * height)),
                        constants.FONT, 1.5, constants.BLUE, 5)

          # Detect Movement
          if moving(landmark):
            cv2.rectangle(frame, (int(0.025 * width), int(0.025 * height)),
                          (int(0.45 * width), int(0.14 * height)), constants.BLACK, -1)
            cv2.putText(frame, "Moving Hand...", (int(0.04 * width), int(0.1 * height)),
                        constants.FONT, 1, constants.WHITE, 2)

        # Display fingers counted
        text = str(finger_count).zfill(2)
        cv2.rectangle(frame, (int(0.025 * width), int(0.825 * height)),
                      (int(0.175 * width), int(0.975 * height)), constants.WHITE, -1)
        cv2.putText(frame, text, (int(0.04 * width), int(0.95 * height)),
                    constants.FONT, 2, constants.RED, 10)
      # Show Frame
      cv2.imshow(constants.TITLE, frame)

      # Until Esc Key is pressed
      if cv2.waitKey(5) & 0xFF == 27:
        break
  capture.release()

if __name__ == "__main__":
  main()


modules/gestures.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Apr 6 2023

@author: iluzioDev

This script implements the gestures and functions used in the Hand Recognition
"""
import numpy

import modules.constants as constants

def distance(p1, p2):
  """
  Calculates the distance between two points on a 2D plane.

  Args:
    p1: Coordinate of the first point
    p2: Coordinate of the second point

  Returns:
    float: Distance between the two points
  """
  return numpy.sqrt(((p2.x - p1.x)**2) + ((p2.y - p1.y)**2))

def sideways(landmark, label):
  """
  Checks if the hand is sideways (Horizontal Aligned)

  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)

  Returns:
    bool: True if hand is sideways, False otherwise
  """
  wrist = landmark[0]
  thumb_cmc = landmark[1]
  pinky_mcp = landmark[17]

  if thumb_cmc.y > wrist.y or thumb_cmc.y < pinky_mcp.y:
    return True
  return False

def raised(landmark, label, tip):
  """
  Checks if a finger is up or not
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
    tip: ID of the tip of the finger
  
  Returns:
    bool: True if finger is up, False otherwise
  """
  # Vertical Aligned Hand Case
  if not sideways(landmark, label):
    # In thumb case, it must be checked if hand is backward or forward
    if tip == constants.TIPS[0]:
      if landmark[tip].x < landmark[constants.TIPS[4]].x:
        if (label == "Left" and landmark[tip].x < landmark[tip - 1].x) or (label == "Right" and landmark[tip].x < landmark[tip - 1].x):
          return True
      if landmark[tip].x > landmark[constants.TIPS[4]].x:
        if (label == "Right" and landmark[tip].x > landmark[tip - 1].x) or (label == "Left" and landmark[tip].x > landmark[tip - 1].x):
          return True
      return False
    # For rest of fingers, make the comparison between tip and pip
    return landmark[tip].y < landmark[tip - 2].y
  
  # Horizontal Aligned Hand Case
  # In thumb case, it must be checked if hand is backward or forward
  if tip == constants.TIPS[0]:
    thumb_cmc = landmark[constants.TIPS[0]]
    pinky_mcp = landmark[17]
    if thumb_cmc.y < pinky_mcp.y:
      return landmark[tip].y < landmark[tip - 1].y
    return landmark[tip].y > landmark[tip - 1].y
  # For rest of fingers, make the comparison between tip and pip
  return (label == "Left" and landmark[tip].x > landmark[tip - 2].x) or (label == "Right" and landmark[tip].x < landmark[tip - 2].x)


def thumb_up(landmark, label):
  """
  Checks if the thumb is up
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
    
  Returns:
    bool: True if thumb is up, False otherwise
  """
  if not sideways(landmark, label):
    return False
  if not raised(landmark, label, constants.TIPS[0]) or (landmark[constants.TIPS[0]].y > landmark[constants.TIPS[0] - 1].y):
    return False
  for i in range(1, 5):
    if raised(landmark, label, constants.TIPS[i]):
      return False
  return True

def thumb_down(landmark, label):
  """
  Checks if the thumb is down
  
  Args: 
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
  
  Returns:
    bool: True if thumb is down, False otherwise
  """
  if not sideways(landmark, label):
    return False
  if not raised(landmark, label, constants.TIPS[0]) or (landmark[constants.TIPS[0]].y < landmark[constants.TIPS[0] - 1].y):
    return False
  for i in range(1, 5):
    if raised(landmark, label, constants.TIPS[i]):
      return False
  return True

def peace(landmark, label):
  """
  Checks if the hand is doing a Peace Sign
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)
    
  Returns:
    bool: True if hand is doing a Peace Sign, False otherwise
  """
  if sideways(landmark, label):
    return False
  if not raised(landmark, label, constants.TIPS[1]) or not raised(landmark, label, constants.TIPS[2]):
    return False
  for i in range(5):
    if i == 1 or i == 2:
      continue
    if raised(landmark, label, constants.TIPS[i]):
      return False
  return True

def rock_n_roll(landmark, label):
  """
  Checks if the hand is doing a Rock N Roll Sign
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)

  Returns:
    bool: True if hand is doing a Rock N Roll Sign, False otherwise
  """
  if not sideways(landmark, label):
    if not raised(landmark, label, constants.TIPS[0]) or not raised(landmark, label, constants.TIPS[1]) or not raised(landmark, label, constants.TIPS[4]):
      return False
    for i in range(2, 4):
      if raised(landmark, label, constants.TIPS[i]):
        return False
  else:
    return False
  return True

def surf(landmark, label):
  """
  Checks if the hand is doing a Surf Sign

  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)

  Returns:
    bool: True if hand is doing a Surf Sign, False otherwise
  """
  if not sideways(landmark, label):
    return False
  if not raised(landmark, label, constants.TIPS[0]) or not raised(landmark, label, constants.TIPS[4]) or (landmark[constants.TIPS[0]].y > landmark[constants.TIPS[0] - 1].y):
    return False
  for i in range(1, 4):
    if raised(landmark, label, constants.TIPS[i]):
      return False
  return True

def ok(landmark, label):
  """
  Checks if the hand is doing an OK Sign
  
  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)

  Returns:
    bool: True if hand is doing an OK Sign, False otherwise
  """
  if sideways(landmark, label):
    return False
  for i in range(2, 5):
    if not raised(landmark, label, constants.TIPS[i]):
      return False
  d = distance(landmark[constants.TIPS[0]], landmark[constants.TIPS[1]])
  if d > constants.MOE:
    return False
  return True

def loser(landmark, label):
  """
  Checks if the hand is doing a Loser Sign

  Args:
    landmark: Landmark of one hand
    label: Label of the hand (Left or Right)

  Returns:
    bool: True if hand is doing a Loser Sign, False otherwise
  """
  if sideways(landmark, label):
    return False
  if not raised(landmark, label, constants.TIPS[0]) or not raised(landmark, label, constants.TIPS[1]):
    return False
  for i in range(2, 5):
    if raised(landmark, label, constants.TIPS[i]):
      return False
  return True


src/main.py
import cv2
import numpy as np
import mediapipe as mp

# Title of the Window with the Hand Recognition
TITLE = 'Handy Hands'

# Margin of Error used in checking if two fingers are touching each other
MARGIN_OF_ERROR = 0.05

# Margin of Hand movement
MOVEMENT_MARGIN = 0.025

# Flags to detect Hand movement
moving = prev_thumb = prev_index = prev_middle = prev_ring = prev_pinky = False

# Colors used
BLACK = (0, 0, 0)
RED = (0, 0, 255)
BLUE = (255, 0, 0)
WHITE = (255, 255, 255)

# Fonts used
FONT = cv2.FONT_HERSHEY_COMPLEX

mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles
mp_hands = mp.solutions.hands

# list of finger tips locators, 4 is thumb, 20 is pinky finger
TIPIDS = [4, 8, 12, 16, 20]

# Calculates the distance between 2 point in a landmark
def distance(p1, p2):
  return np.sqrt(((p2.x - p1.x)**2) + ((p2.y - p1.y)**2))

# Check if hand is sideways (Horizontal Aligned)
def isHandSideways(landmark, label):
  wrist = landmark[0]
  thumb_cmc = landmark[1]
  pinky_mcp = landmark[17]
  if thumb_cmc.y > wrist.y or thumb_cmc.y < pinky_mcp.y:
    return True
  return False

# Checks if finger is up or not
def isFingerUp(landmark, label, tip):
  # Vertical Aligned Hand Case
  if not isHandSideways(landmark, label):
    # In thumb case, it must be checked if hand is backward or forward
    if tip == TIPIDS[0]:
      if landmark[tip].x < landmark[TIPIDS[4]].x:
        if (label == "Left" and landmark[tip].x < landmark[tip - 1].x) or (label == "Right" and landmark[tip].x < landmark[tip - 1].x):
          return True
      if landmark[tip].x > landmark[TIPIDS[4]].x:
        if (label == "Right" and landmark[tip].x > landmark[tip - 1].x) or (label == "Left" and landmark[tip].x > landmark[tip - 1].x):
          return True
      return False
    # For rest of fingers, make the comparison between tip and pip
    return landmark[tip].y < landmark[tip - 2].y
  # Horizontal Aligned Hand Case
  else:
    # In thumb case, it must be checked if hand is backward or forward
    if tip == TIPIDS[0]:
      thumb_cmc = landmark[TIPIDS[0]]
      pinky_mcp = landmark[17]
      if thumb_cmc.y < pinky_mcp.y:
        return landmark[tip].y < landmark[tip - 1].y
      return landmark[tip].y > landmark[tip - 1].y
    # For rest of fingers, make the comparison between tip and pip
    return (label == "Left" and landmark[tip].x > landmark[tip - 2].x) or (label == "Right" and landmark[tip].x < landmark[tip - 2].x)
  return False

# Counts fingers up in a hand
def countFingers(landmark, label):
  count = 0
  for i in range(5):
    if isFingerUp(landmark, label, TIPIDS[i]):
      count += 1
  return count

# Detect a Thumbs Up Gesture
def detectThumbUp(landmark, label):
  if isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[0]) or (landmark[TIPIDS[0]].y > landmark[TIPIDS[0] - 1].y):
      return False
    for i in range(1, 5):
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detect a Thumbs Down Gesture
def detectThumbDown(landmark, label):
  if isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[0]) or (landmark[TIPIDS[0]].y < landmark[TIPIDS[0] - 1].y):
      return False
    for i in range(1, 5):
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detect a Peace Sign Gesture
def detectPeaceSign(landmark, label):
  if not isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[1]) or not isFingerUp(landmark, label, TIPIDS[2]):
      return False
    for i in range(5):
      if i == 1 or i == 2:
        continue
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detect a Rock N Roll Sign Gesture
def detectRockNRollSign(landmark, label):
  if not isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[0]) or not isFingerUp(landmark, label, TIPIDS[1]) or not isFingerUp(landmark, label, TIPIDS[4]):
      return False
    for i in range(2, 4):
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detect a Surf Sign Gesture
def detectSurfSign(landmark, label):
  if isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[0]) or not isFingerUp(landmark, label, TIPIDS[4]) or (landmark[TIPIDS[0]].y > landmark[TIPIDS[0] - 1].y):
      return False
    for i in range(1, 4):
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detect a Ok Sign Gesture
def detectOkSign(landmark, label):
  if not isHandSideways(landmark, label):
    for i in range(2, 5):
      if not isFingerUp(landmark, label, TIPIDS[i]):
        return False
    d = distance(landmark[TIPIDS[0]], landmark[TIPIDS[1]])
    if d > MARGIN_OF_ERROR:
      return False
  else:
    return False
  return True

# Detect a Loser Sign (L) Gesture
def detectLoserSign(landmark, label):
  if not isHandSideways(landmark, label):
    if not isFingerUp(landmark, label, TIPIDS[0]) or not isFingerUp(landmark, label, TIPIDS[1]):
      return False
    for i in range(2, 5):
      if isFingerUp(landmark, label, TIPIDS[i]):
        return False
  else:
    return False
  return True

# Detects gestures in a hand
def detectGestures(landmark, label):
  if detectThumbUp(landmark, label):
    return "Thumbs Up! :D"
  if detectThumbDown(landmark, label):
    return "Thumbs Down... :("
  if detectPeaceSign(landmark, label):
    return "Peace!"
  if detectRockNRollSign(landmark, label):
    return "Rock'n'Roll!"
  if detectSurfSign(landmark, label):
    return "Surf's up!"
  if detectOkSign(landmark, label):
    return "Everything is OK"
  if detectLoserSign(landmark, label):
    return "Loser >:D"
  return False

# Detects Hand Movements calculating the distance in a margin between the positions
# of 2 instances through time of the hand
def detectMovement(landmark):
  global moving, prev_thumb, prev_index, prev_middle, prev_ring, prev_pinky
  if not prev_thumb:
    prev_thumb  = landmark[TIPIDS[0]]
    prev_index  = landmark[TIPIDS[1]]
    prev_middle = landmark[TIPIDS[2]]
    prev_ring   = landmark[TIPIDS[3]]
    prev_pinky  = landmark[TIPIDS[4]]

  d0 = distance(prev_thumb, landmark[TIPIDS[0]])
  d1 = distance(prev_index, landmark[TIPIDS[1]])
  d2 = distance(prev_middle, landmark[TIPIDS[2]])
  d3 = distance(prev_ring, landmark[TIPIDS[3]])
  d4 = distance(prev_pinky, landmark[TIPIDS[4]])

  if (d0 + d1 + d2 + d3 + d4) > MOVEMENT_MARGIN:
    moving = True
  else:
    moving = False

  prev_thumb  = landmark[TIPIDS[0]]
  prev_index  = landmark[TIPIDS[1]]
  prev_middle = landmark[TIPIDS[2]]
  prev_ring   = landmark[TIPIDS[3]]
  prev_pinky  = landmark[TIPIDS[4]]

  return moving

# Starts the WebCam Capture
capture = cv2.VideoCapture(0)
with mp_hands.Hands(model_complexity=0, min_detection_confidence=0.5, min_tracking_confidence=0.5) as hands:
  while capture.isOpened():
    success, frame = capture.read()
    if not success:
      print("Ignoring empty camera frame.")
      # If loading a video, use 'break' instead of 'continue'.
      continue

    # Flip the frame horizontally for a selfie-view display.
    frame = cv2.flip(frame, 1)

    # To improve performance, optionally mark the frame as not writeable to
    # pass by reference.
    frame.flags.writeable = False
    frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    results = hands.process(frame)

    # Draw the hand annotations on the frame.
    frame.flags.writeable = True
    frame = cv2.cvtColor(frame, cv2.COLOR_RGB2BGR)

    # height, width and depth (RGB=3) of frame
    (height, width, depth) = frame.shape

    # finger_count of raised fingers
    finger_count = 0

    # If hands are detected
    if results.multi_hand_landmarks:
      for hand_landmarks in results.multi_hand_landmarks:
        # Index Position and Handedness (left hand or rigth hand)
        index = results.multi_hand_landmarks.index(hand_landmarks)
        label = results.multi_handedness[index].classification[0].label

        # Drawing within frame of the hand landmark with its connections and mediapipe default style
        mp_drawing.draw_landmarks(
            frame,
            hand_landmarks,
            mp_hands.HAND_CONNECTIONS,
            mp_drawing_styles.get_default_hand_landmarks_style(),
            mp_drawing_styles.get_default_hand_connections_style())

        # Detect fingers
        landmark = hand_landmarks.landmark
        finger_count += countFingers(landmark, label)

        # Detect Gestures
        if detectGestures(landmark, label):
          cv2.rectangle(frame, (int(0.2 * width), int(0.825 * height)),
                        (int(0.975 * width), int(0.975 * height)), WHITE, -1)
          cv2.putText(frame, detectGestures(landmark, label), (int(0.225 * width), int(0.925 * height)),
                      FONT, 1.5, BLUE, 5)

        # Detect Movement
        if detectMovement(landmark):
          cv2.rectangle(frame, (int(0.025 * width), int(0.025 * height)),
                        (int(0.45 * width), int(0.14 * height)), BLACK, -1)
          cv2.putText(frame, "Moving Hand...", (int(0.04 * width), int(0.1 * height)),
                      FONT, 1, WHITE, 2)

      # Display fingers counted
      text = str(finger_count).zfill(2)
      cv2.rectangle(frame, (int(0.025 * width), int(0.825 * height)),
                    (int(0.175 * width), int(0.975 * height)), WHITE, -1)
      cv2.putText(frame, text, (int(0.04 * width), int(0.95 * height)),
                  FONT, 2, RED, 10)
    # Show Frame
    cv2.imshow(TITLE, frame)

    # Until Esc Key is pressed
    if cv2.waitKey(5) & 0xFF == 27:
      break
capture.release()
